pennylane/plugins/default_qubit.py

```diff
@@ -309,79 +309,82 @@ class DefaultQubit(QubitDevice):
 
-        n = self.num_wires
+    def apply(self, operations, rotations=None):
+        rotations = rotations or []
 
-        if isinstance(operation, QubitStateVector):
-            input_state = np.asarray(par[0], dtype=np.complex128)
+        # apply the circuit operations
+        for i, operation in enumerate(operations):
+            # number of wires on device
+            wires = operation.wires
+            par = operation.parameters
 
-            if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
-                raise ValueError("Sum of amplitudes-squared does not equal one.")
-            n_state_vector = input_state.shape[0]
-            if not self._first_operation:
+            if i > 0 and isinstance(operation, (QubitStateVector, BasisState)):
                 raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
                                   "on a {} device.".format(operation.name, self.short_name))
-            if input_state.ndim == 1 and n_state_vector == 2**len(wires):
 
-                # generate basis states on subset of qubits via the cartesian product
-                tuples = np.array(list(itertools.product([0, 1], repeat=len(wires))))
+            if isinstance(operation, QubitStateVector):
+                self.apply_state_vector(wires, par)
 
-                # get basis states to alter on full set of qubits
-                unravelled_nums = np.zeros((2 ** len(wires), n), dtype=int)
-                unravelled_nums[:, wires] = tuples
+            elif isinstance(operation, BasisState):
+                self.apply_basis_state(wires, par)
 
-                # get indices for which the state is changed to input state vector elements
-                nums = np.ravel_multi_index(unravelled_nums.T, [2] * n)
-                self._state = np.zeros_like(self._state)
-                self._state[nums] = input_state
             else:
-                raise ValueError("State vector must be of length 2**wires.")
-            self._first_operation = False
-            return
-
-        if isinstance(operation, BasisState):
-            # length of basis state parameter
-            n_basis_state = len(par[0])
-
-            if not set(par[0]).issubset({0, 1}):
-                raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
-            if n_basis_state != len(wires):
-                raise ValueError("BasisState parameter and wires must be of equal length.")
-            if not self._first_operation:
-                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
-                                  "on a {} device.".format(operation.name, self.short_name))
+                A = self._get_operator_matrix(operation.name, par)
+                self._state = self.mat_vec_product(A, self._state, wires)
+
+        # store the pre-rotated state
+        self.state = self._state
```

Comment:
It would be nice to store it in an attribute such that it will be easy to read later, such as ``_pre_rotated_state`` or something similar, as currently the user would need to bear in mind the distinction between ``_state`` and ``state`` at all times

Reply:
I agree. Noe that `device.state` should be considered an official part of the API that all state-based simulators should support, whereas `_state` is not part of the API and its use is left up to the plugin.

Ideally, we should have this:
```python
@property
def state(self):
    """Returns the state vector of the circuit prior to measurement.

    .. note::

        Only state vector simulators support this property. Please see the
        plugin documentation for more details.
    """
    return self._pre_rotated_state
```

Reply:
Yes so I find myself confused about this at times, but this makes it clearer, thanks!

So there would be currently three different types of states that might be of interest to a user:
1) *pre-rotated state*: state obtained by applying the series of unitaries contained in the quantum circuit on the quantum state obtained after state preparation
2) *rotated state*: the state that is obtained by rotating the ``pre-rotated state`` based on the specified observable
3) *post-measurement state*: the state that is obtained by measuring the ``rotated_state`` on each wire in the ``PauliZ`` basis (this is after the collapse)

And so currently ``state`` would correspond to returning 1).

Reply:
Yep. I don't see a use case for supporting (2) (I imagine most users won't realise the rotation is even happening), and (3) corresponds to `return qml.sample()`, I think?

Reply:
I think now ``_state`` will contain (2) actually (as per the rotations being carried out on lines 344-348) and (3) could be recreated if in ``QubitDevice.generate_samples`` stored the post-measurement states as well. That is, currently ``QubitDevice.generate_samples`` generates samples based on the eigenvalues, but that is done through sampling a state as well (which is not preserved).

This goes back to the idea of storing not just the samples but the post-measurement states as well. However, for ``shots=1000`` the naive method would be storing ``1000`` states as well, which can get really memory-intensive when stored. A bit of a better way would be storing ``len(wires)`` long bitstring of the basis state that was sampled as a tuple. Then, in a dictionary, we could have the ``2 ** len(wires)`` many basis states as keys and the post-measurement states as values, and recreating the post-measurement state could come by returning the state corresponding to the sampled basis state.

Reply:
Btw. provided there is no main flaw in this latter logic, this could be used to have gates conditioned on measurement outcomes in PL

Reply:
(But for larger systems, e.g. >20 qubits the size of such dictionaries would also be pretty big)

Reply:
Ah, I see what you mean by post-measurement state! This is definitely something to consider in the future --- for now, the design of the API is very much suited for near term quantum devices, where feed-forward measurement results are not supported.

Reply:
Have added in the abstract `state` property!

Reply:
Yep, I agree!